src/chains/ethereum/ethereum/src/blockchain.ts

```diff
@@ -1274,42 +1266,49 @@ export default class Blockchain extends Emittery.Typed<
         Buffer /*codeHash*/
       ])[2];
 
-      let keys: Buffer[] = [];
-      return new Promise((resolve, reject) => {
-        storageTrie
-          .createReadStream()
-          .on("data", data => {
-            keys.push(data.key);
-          })
-          .on("end", () => {
-            // #4 - sort and filter keys
-            const sortedKeys = keys.sort((a, b) => Buffer.compare(a, b));
-
-            // find starting point in array of sorted keys
-            const startKeyBuffer = Data.from(startKey).toBuffer();
-            keys = sortedKeys.filter(key => {
-              if (Buffer.compare(startKeyBuffer, key) <= 0) {
-                return key;
-              }
-            });
+      return new Promise<T>((resolve, reject) => {
+        const startKeyBuffer = Data.from(startKey).toBuffer();
+        const compare = (a: Buffer, b: Buffer) => a.compare(b) < 0;
+
+        const keys: Buffer[] = [];
+        const handleData = ({ key }) => {
+          // ignore anything that comes before our starting point
+          if (startKeyBuffer.compare(key) > 0) return;
+
+          // #4 - sort and filter keys
+          // insert the key exactly where it needs to go in the array
+          const position = findInsertPosition(keys, key, compare);
+          // ignore if the value couldn't possibly be relevant
+          if (position > maxResult) return;
+          keys.splice(position, 0, key);
+        };
 
-            // only take the maximum number of entries requested
-            keys = keys.slice(0, maxResult + 1);
-            if (keys.length > maxResult) {
-              // assign nextKey and remove it from array of keys
-              const nextKey = keys.pop();
-              result.nextKey = Data.from(nextKey).toJSON();
-            }
+        const handleEnd = () => {
+          if (keys.length > maxResult) {
+            // we collected too much data, so we've got to trim it a bit
+            resolve({
+              // only take the maximum number of entries requested
+              keys: keys.slice(0, maxResult),
+              // assign nextKey
+              nextKey: Data.from(keys[maxResult])
+            });
+          } else {
+            resolve({
+              keys,
+              nextKey: null
+            });
+          }
+        };
 
-            resolve(keys);
-          });
+        const rs = storageTrie.createReadStream();
```

Comment:
We can't get keys to come back already ordered because the merkle patricia tree library we use parallelizes up to 500 database requests at once.

This just means we have to sort client side, which could mean performance takes a hit.

But generally I think `debug_storageRangeAt` will attempt to consume all available storage in one go, which means we would have had to load all the data anyway.

Where we'd take a hit is when we only need to return a small amount of storage. In that case we still have to load all of it.

I think turning the priority queue it uses internally into a FIFO queue would actually be very complicated, because it resolves both branches and leaves, and it'd have to ensure the proper ordering there. Not positive though.

Since that's not going to happen any time soon, the next best thing is to just insert them in order as we get them, instead of a big oll sort at the end.

This will suck a bit more is after we get the in-memory cache in place. We won't need any parallelization in that case; and it'll actually just slow everything down (it'll still be faster than hitting leveldb, of course).

Thanks for watching. Don't forget to smash that like button!